### PR TITLE
MelloTrainer v0.4.0

### PR DESCRIPTION
**Pull request by WolfKnight98**
New vehicle and skin saving system.

### CODE REVIEW

File: mellotrainer/utils.lua [hunk 1]
```diff
@@ -1,23 +1,30 @@
...
-		RequestClipSet( set )
-		Citizen.Wait( 5 )
-	end 
+    while ( not HasClipSetLoaded( set ) ) do 
+        RequestClipSet( set )
+        Citizen.Wait( 5 )
+    end 
+end 
+
+function _LoadModel( mdl )
```
Comment (TheStonedTurtle) [inline mellotrainer/utils.lua]: I thought this was causing issues, why are we keeping this function in?

File: mellotrainer/utils.lua [hunk 2]
```diff
@@ -79,22 +86,35 @@ function GetToggleState(variableName)
   end
 end
 
+-- Removes invalid characters and spaces from a string, for use
+-- when storing information using DATASAVE. 
+function EscapeStringData( str )
```
Comment (TheStonedTurtle) [inline mellotrainer/utils.lua]: Is this still being used anywhere?

File: mellotrainer/vehicles.lua [hunk 3]
```diff
@@ -18,13 +18,13 @@ local function SpawnVehicle(model, x, y, z, heading, ped)
...
 	if IsModelValid(model) then
-		RequestModel(model)
-		while not HasModelLoaded(model) do
-			Wait(1)
-		end
-
-		local veh = CreateVehicle(model, x, y, z + 1, heading, true, true)
+		RequestModel( model )
+		while ( not HasModelLoaded( model ) ) do 
+			RequestModel( model )
```
Comment (TheStonedTurtle) [inline mellotrainer/vehicles.lua]: You shouldn't need to re-request the model if it hasn't loaded.

File: mellotrainer/vehicles.lua [hunk 4]
```diff
@@ -18,13 +18,13 @@ local function SpawnVehicle(model, x, y, z, heading, ped)
...
+			Citizen.Wait( 0 )
+		end 
 
+		local veh = CreateVehicle( model, x, y, z + 1, heading, true, true )
```
Comment (TheStonedTurtle) [inline mellotrainer/vehicles.lua]: You forgot to revert this change to spawning in front not on top.

File: mellotrainer/vehicles.lua [hunk 5]
```diff
@@ -75,7 +77,14 @@ end)
 
 RegisterNUICallback("vehspawn", function(data, cb)
 	local playerPed = GetPlayerPed(-1)
-	local x, y, z = table.unpack(GetOffsetFromEntityInWorldCoords(playerPed, 0.0, 7.5, 0.0))
+	local x, y, z
+
+	if ( featureSpawnInsideCar ) then
```
Comment (TheStonedTurtle) [inline mellotrainer/vehicles.lua]: Think this should be changed to be a spawn setting and not default functionality. All vehicles should spawn the same regardless of `featureSpawnInsideCar`. There should be a `featureSpawnCarAtPlayerPosition` or something similar to determine this.

File: mellotrainer/vehicles.lua [hunk 6]
```diff
@@ -90,11 +99,396 @@ RegisterNUICallback("vehspawn", function(data, cb)
...
+end )
+
+RegisterNUICallback( "vehiclesave", function( data, cb )
+    Citizen.CreateThread( function() 
+    	local ped = GetPlayerPed( -1 )
+    	local vehicleTableData = {}
+        local saveName = nil 
+        local overwriting 
+        local renaming 
+        local index
```
Comment (TheStonedTurtle) [inline mellotrainer/vehicles.lua]: nice spacing.

File: mellotrainer/vehicles.lua [hunk 7]
```diff
@@ -90,11 +99,396 @@ RegisterNUICallback("vehspawn", function(data, cb)
...
+    	if ( DoesEntityExist( ped ) and not IsEntityDead( ped ) ) then 
+    		local veh = GetVehiclePedIsIn( ped, false )
+
+    		if ( GetPedInVehicleSeat( veh, -1 ) == ped ) then 
+                while ( (saveName == nil and not overwriting) or (saveName == nil and renaming) ) do 
+                    saveName = requestInput( "Enter save name", 24 )
+                    Citizen.Wait( 1 )
+                end 
+
+                if ( saveName or overwriting or renaming ) then
```
Comment (TheStonedTurtle) [inline mellotrainer/vehicles.lua]: What happened with the spacing below here, too many indents?

File: mellotrainer/vehicles.lua [hunk 8]
```diff
@@ -90,11 +99,396 @@ RegisterNUICallback("vehspawn", function(data, cb)
...
+					end 
+
+					vehicleTableData[ "neonColour" ] = { neonR, neonG, neonB }
+					vehicleTableData[ "smokeColour" ] = { smokeR, smokeG, smokeB }
+					vehicleTableData[ "neonToggles" ] = neonToggles
+
+					local extras = {}
+
+                    if ( DoesVehicleHaveExtras( veh ) ) then 
+            			for i = 1, 30 do
```
Comment (TheStonedTurtle) [inline mellotrainer/vehicles.lua]: space

File: mellotrainer/data-saving.lua [hunk 9]
```diff
@@ -0,0 +1,184 @@
+os.execute( "mkdir mtsaves" )
+
+DATASAVE = {}
+
+DATASAVE.dir = "mtsaves/"
+DATASAVE.players = {}
+
+function DATASAVE:GetSteamId( source )
```
Comment (WolfKnight98) [inline mellotrainer/data-saving.lua]: I think it'll be fine in here for now, plus, the utilities file is clientsided.

File: mellotrainer/data-saving.lua [hunk 10]
```diff
@@ -0,0 +1,184 @@
+os.execute( "mkdir mtsaves" )
```
Comment (WolfKnight98) [inline mellotrainer/data-saving.lua]: I'll work on something.

File: Changelog.md [hunk 11]
```diff
@@ -1,3 +1,22 @@
...
+    <li>Fixed being able to open the trainer in the pause menu</li>
+    <li>Fixed some vehicle modification names</li>
+    <li>Potential fix for god mode not working</li>
+    <li>Adjusted order of some NUI tables</li>
+</ul>
+
+<h3>Additions</h3>
+<ul>
+	<li>New data-saving.lua file for serversided file storage and manipulation</li>
+	<li>New vehicle and skin saving system, this is in beta right now. In order for the saving system to work players must be logged in to Steam.</li>
```
Comment (WolfKnight98) [inline Changelog.md]: Will do.

File: mellotrainer/data-saving.lua [hunk 12]
```diff
@@ -0,0 +1,184 @@
...
+
+        if ( next( skinData ) ~= nil ) then 
+            TriggerClientEvent( 'wk:RecieveSavedSkins', source, skinData )
+        end 
+    else 
+        RconPrint( "MELLOTRAINER: Attempted to load save data for " .. GetPlayerName( source ) .. ", but does not have a steam id.\n" ) 
+    end 
+end 
+
+function DATASAVE:GetSteamIdFromSource( source )
```
Comment (WolfKnight98) [inline mellotrainer/data-saving.lua]: I've removed the serverside source->steamid storage.

File: mellotrainer/data-saving.lua [hunk 13]
```diff
@@ -0,0 +1,184 @@
...
+    else 
+        DATASAVE.players[source] = nil 
+        RconPrint( "MELLOTRAINER: " .. GetPlayerName( source ) .. " is not connecting with a steam id.\nPlayer will not have the ability to save/load.\n" )
+    end 
+end )
+
+RegisterServerEvent( 'wk:DataSave' )
+AddEventHandler( 'wk:DataSave', function( type, data, index )
+    RconPrint( "Got wk:DataSave from " .. GetPlayerName( source ) .. " " .. source .. "\n"  )
+    RconPrint( "SteamID: " .. DATASAVE.players[source] .. "\n" )
```
Comment (WolfKnight98) [inline mellotrainer/data-saving.lua]: I've removed all of the debug messages.

File: mellotrainer/data-saving.lua [hunk 14]
```diff
@@ -0,0 +1,184 @@
...
+
+    if ( id ~= nil ) then 
+        local file = id .. "_" .. type .. ".txt"
+        DATASAVE:WriteToFile( file, data, index )
+    else 
+        RconPrint( "MELLOTRAINER: " .. GetPlayerName( source ) .. " attempted to save, but does not have a steam id.\n" ) 
+    end 
+end )
+
+AddEventHandler( 'playerDropped', function()
```
Comment (WolfKnight98) [inline mellotrainer/data-saving.lua]: See above messages.

File: mellotrainer/utils.lua (hunk 1 above)
Comment (WolfKnight98) [inline mellotrainer/utils.lua]: I got it to work now.

File: mellotrainer/player-skin.lua [hunk 15]
```diff
@@ -109,7 +349,7 @@ RegisterNUICallback("playerpropmodify", function(data, cb)
 		local validComponents = checkValidPropComponents(propComponents[i].t)
 
 		if #validComponents > 0 then
-			table.insert(validOptions, {
+			table.insert(validOptions, 1, {
```
Comment (WolfKnight98) [inline mellotrainer/player-skin.lua]: It reverses the list.

File: mellotrainer/player-skin.lua [hunk 16]
```diff
@@ -72,7 +312,7 @@ RegisterNUICallback("playerskinmodify", function(data, cb)
 		local validComponents = checkValidComponents(components[i].t)
 
 		if #validComponents > 0 then
-			table.insert(validOptions, {
+			table.insert(validOptions, 1, {
```
Comment (WolfKnight98) [inline mellotrainer/player-skin.lua]: It reverses the list.

File: mellotrainer/utils.lua (hunk 2 above)
Comment (WolfKnight98) [inline mellotrainer/utils.lua]: I've removed the function.

File: mellotrainer/vehicles.lua (hunk 3 above)
Comment (WolfKnight98) [inline mellotrainer/vehicles.lua]: Makes it load the model a tiny bit faster.

File: mellotrainer/data-saving.lua (hunk 9 above)
Comment (TheStonedTurtle) [inline mellotrainer/data-saving.lua]: Good point.
